In [4]:
# Importing required packages

import statsmodels.api as sm
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm          # A normal continuous random variable
from numpy.random import seed         # set the randomness
from scipy import stats               # statistical operations
import warnings
warnings.simplefilter("ignore")

# Hypothesis Testing

Let's create some dummy age data for the population of voters in the entire country and a sample of voters in North_Carolina and test whether the average age of voters in North_Carolina differs from the entire country population.

.rvs provides random samples; scale = std. deviation, loc = mean, size = no. of samples

In [7]:
# Generate Population ages data

np.random.seed(6)

population_ages1 = stats.norm.rvs(scale=18, loc=45, size=150000)
population_ages2 = stats.norm.rvs(scale=18, loc=20, size=100000)
population_ages = np.concatenate((population_ages1, population_ages2))

# Generate North Carolina sample ages data

North_Carolina_ages1 = stats.norm.rvs(scale=18, loc=40, size=30)
North_Carolina_ages2 = stats.norm.rvs(scale=18, loc=20, size=25)
North_Carolina_ages = np.concatenate((North_Carolina_ages1, North_Carolina_ages2))

print("Mean age of population of voters in the entire country: ", population_ages.mean())
print("Mean age of North Carolina population: ", North_Carolina_ages.mean())

Mean age of population of voters in the entire country:  34.97043368335889
Mean age of North Carolina population:  28.52661116077034


Notice that we used a slightly different combination of distributions to generate the sample data for North_Carolina, so we know that the two means are different. Let's conduct a z-test at a 95% confidence level and see if it correctly rejects the null hypothesis that the sample comes from the same distribution as the population.

In [11]:
# Perform Z-test for mean
one_sample_ztest = sm.stats.ztest(x1= North_Carolina_ages, x2=None, value= population_ages.mean(), alternative='smaller')

print("test-statistic (z-score): ", one_sample_ztest[0])
print("p-value: ", one_sample_ztest[1])

test-statistic (z-score):  -2.2814904209459996
p-value:  0.011259721451099942


The z-score tells us how many standard deviations from the mean our score is. In this example, our score is -2.28 standard deviations below the mean. Also, we can reject the null hypothesis that the mean of samples is equal to the population mean (i.e. 28.52) as we can see that the p-value is less than 0.05 (i.e. 0.011). This indicates that the difference between the sample mean and population mean is statistically significant.